File: js/src/providers/gemini.js
// Google Gemini adapter — mirror of gemini.py.
import { HttpError, requestJson } from "../http.js";

export class GeminiProvider {
  constructor({ model, baseUrl = null, keyEnv = null, apiKey = null, timeout = 90000, maxRetries = 4 }) {
    this.model = model.startsWith("models/") ? model.slice("models/".length) : model;
    this.base = (baseUrl || "https://generativelanguage.googleapis.com/v1beta").replace(/\/+$/, "");
    this.keyEnv = keyEnv || "GEMINI_API_KEY";
    this.timeout = timeout;
    this.maxRetries = maxRetries;
    this.apiKey = apiKey ?? process.env[this.keyEnv] ?? process.env.GOOGLE_API_KEY;
    if (!this.apiKey) throw new Error(`set ${this.keyEnv} to use the gemini provider (model '${model}')`);
  }

  body(messages, sampling) {
    const contents = [];
    const system = [];
    for (const m of messages) {
      if (m.role === "system") { system.push(m.content); continue; }
      contents.push({ role: m.role === "assistant" ? "model" : "user", parts: [{ text: m.content }] });
    }
    const cfg = { temperature: sampling.temperature, maxOutputTokens: sampling.max_tokens };
    if (sampling.top_p !== 1.0) cfg.topP = sampling.top_p;
    const b = { contents, generationConfig: cfg };
    if (system.length) b.systemInstruction = { parts: [{ text: system.join(" ") }] };
    return b;
  }

  async chat(messages, sampling) {
    const url = `${this.base}/models/${this.model}:generateContent`;
    const headers = { "x-goog-api-key": this.apiKey };
    const body = await requestJson(url, { headers, body: this.body(messages, sampling), timeout: this.timeout, maxRetries: this.maxRetries });
    const cands = body.candidates;
    if (!cands || !cands.length) {
      const err = body.error;
      throw new HttpError(`no candidates in response: ${(err && err.message) || JSON.stringify(body).slice(0, 300)}`, { transient: false });
    }
    const parts = (cands[0].content && cands[0].content.parts) || [];
    const text = parts.filter((p) => "text" in p).map((p) => p.text || "").join("");
    const um = body.usageMetadata || {};
    return {
      text,
      meta: { usage: { promptTokenCount: um.promptTokenCount, candidatesTokenCount: um.candidatesTokenCount, totalTokenCount: um.totalTokenCount }, served_by: "gemini", model_returned: body.modelVersion || this.model, response_id: body.responseId, finish_reason: cands[0].finishReason },
    };
  }
}

File: src/aieasybatch/providers/gemini.py
"""Google Gemini adapter (Generative Language API).

Its own shape again: the model goes in the URL (``/models/<model>:generateContent``),
auth is the ``x-goog-api-key`` header, turns are ``contents:[{role, parts:[{text}]}]``
with the assistant role spelled ``model``, sampling lives under ``generationConfig``,
and the system prompt is ``systemInstruction``.
"""
from __future__ import annotations

import os

from .base import Provider
from .http import HttpError, request_json


class GeminiProvider(Provider):
    name = "gemini"

    def __init__(self, model, *, base_url=None, key_env=None,
                 timeout=90.0, max_retries=4, api_key=None):
        self.model = model.split("/", 1)[1] if model.startswith("models/") else model
        self.base_url = (base_url or "https://generativelanguage.googleapis.com/v1beta").rstrip("/")
        self.key_env = key_env or "GEMINI_API_KEY"
        self.timeout = timeout
        self.max_retries = max_retries
        self.api_key = (api_key or os.environ.get(self.key_env)
                        or os.environ.get("GOOGLE_API_KEY"))
        if not self.api_key:
            raise ValueError(f"set {self.key_env} to use the gemini provider (model {model!r})")

    def _payload(self, messages, sampling):
        contents, system = [], []
        for m in messages:
            if m["role"] == "system":
                system.append(m["content"])
                continue
            role = "model" if m["role"] == "assistant" else "user"
            contents.append({"role": role, "parts": [{"text": m["content"]}]})
        cfg = {"temperature": sampling.temperature, "maxOutputTokens": sampling.max_tokens}
        if sampling.top_p != 1.0:
            cfg["topP"] = sampling.top_p
        body = {"contents": contents, "generationConfig": cfg}
        if system:
            body["systemInstruction"] = {"parts": [{"text": " ".join(system)}]}
        return body

    def chat(self, messages, sampling):
        url = f"{self.base_url}/models/{self.model}:generateContent"
        headers = {"x-goog-api-key": self.api_key, "Content-Type": "application/json"}
        body = request_json(url, headers=headers, body=self._payload(messages, sampling),
                            timeout=self.timeout, max_retries=self.max_retries)
        cands = body.get("candidates")
        if not cands:
            err = body.get("error")
            msg = (err.get("message") if isinstance(err, dict) else err) or str(body)[:300]
            raise HttpError(f"no candidates in response: {msg}", transient=False)
        parts = (cands[0].get("content") or {}).get("parts") or []
        text = "".join(p.get("text", "") for p in parts if "text" in p)
        um = body.get("usageMetadata") or {}
        meta = {
            "usage": {"promptTokenCount": um.get("promptTokenCount"),
                      "candidatesTokenCount": um.get("candidatesTokenCount"),
                      "totalTokenCount": um.get("totalTokenCount")},
            "served_by": "gemini",
            "model_returned": body.get("modelVersion") or self.model,
            "response_id": body.get("responseId"),
            "finish_reason": cands[0].get("finishReason"),
        }
        return text, meta
